docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,694 @@
+          name: backend
+          tags:
+           version: v2
+```
+
+Here, the `/v1` rule from `owner` is unchanged and the `/v2` rule from
+`consumer` overrides that of `owner`.
+
+#### `MeshGateway`
+
+This new resource is intended to replace the `MeshGatewayRoute` resource for
+configuring routes for a `MeshGateway`.
+
+There are two different ways we can express a `MeshGateway` attachment either as
+`spec.targetRef` & `to.targetRef` and `from.targetRef` & `spec.targetRef`.
+
+The core difference is that routes for a `MeshGateway` are fundamentally different from in-mesh
+routes because there's a proxy "in between" requests. That is, when I send a
+request from service to service, the Envoy configuration and routing happens on the source
+proxy side. When I send a request to a `MeshGateway`, the request reaches the
+`MeshGateway` proxy and _then_ gets routed, i.e. they're on the destination
+side.
+
+##### `MeshGateway` as `to.targetRef`
+
+This is a somewhat different way of thinking about `targetRef` because here the
+`to.targetRef` points to the Envoy whose configuration is changing, as opposed
+to `spec.targetRef`. However, a `MeshGateway` proxy is fundamentally different from a
+sidecar proxy because it does not attach to a service, so perhaps this is
+justified?
+
+###### Edge gateway
+
+In order to configure a non-cross-mesh `MeshGateway` resource,
+one option is that the user points `to.targetRef` to a `MeshGateway`.
+
+```yaml
+targetRef:
+ kind: Mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+```
+
+the `spec.targetRef` of `kind: Mesh` can be read as simply meaning "any
+connections made to the `MeshGateway`". Only `kind: Mesh` is permitted with a
+`to.targetRef` of `kind: MeshGateway`.
+
+###### Cross-mesh gateway
+
+For cross-mesh `MeshGateways`:
+
+```yaml
+targetRef:
+ kind: Mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: mesh-gateway
+```
+
+would target _any requests_ from _any `Mesh`_ made to this `MeshGateway` whereas
+`name` must be used to target a specific `Mesh` as source:
+
+```yaml
+targetRef:
+ kind: Mesh
+ name: other-mesh
+to:
+ - targetRef:
+    kind: MeshGateway
+    name: mesh-gateway
+```
+
+##### `MeshGateway` in `spec.targetRef`
+
+If we put `MeshGateway` in `spec.targetRef`, we could instead use `from`.
+
+###### Edge gateway
+
+In order to configure a non-cross-mesh `MeshGateway` resource:
+
+```yaml
+targetRef:
+ kind: MeshGateway
+ name: edge-gateway
+from:
+ - targetRef:
+    kind: Mesh
+```
+
+the `from.targetRef` of `kind: Mesh` can be read as simply meaning "any
+connections made to the `MeshGateway`". Only `kind: Mesh` is permitted as a
+`from.targetRef` with `targetRef` of `MeshGateway`.
+
+###### Cross-mesh gateway
+
+For cross-mesh `MeshGateways`:
+
+```yaml
+targetRef:
+ kind: MeshGateway
+ name: mesh-gateway
+from:
+ - targetRef:
+    kind: Mesh
+```
+
+would target _any requests_ from _any `Mesh`_ made to this `MeshGateway` whereas
```

Comment:
I'm still not convinced by this one. I feel like this backs us in a corner for these reasons:

- For me `kind: Mesh name:""` has always been current Mesh (because when merging with and without a name would have the same priority)
- What if one day I want to identify different non meshed and meshed traffic?

Wouldn't adding `Any` at the risk of deprecating it later would be better?

Reply:
Yeah we can add a new `kind`. For all other cases besides cross mesh gateway, it will be the same as `kind: Mesh`. Which is kind of confusing but maybe not a big deal.

I guess `{kind: Mesh, name: ""}` to me was never _really_ "current Mesh", it's always been just the identity element in our `Dataplane` refinement logic.

> What if one day I want to identify different non meshed and meshed traffic?

Not quite sure what case this means.

Reply:
> What if one day I want to identify different non meshed and meshed traffic?

If one day I want to do:

```yaml
from:
  - targetRef:
      kind: Any # This applies to all traffic regardless of where it's coming from
  - targetRef:
       kind: Non-mesh # This applies to traffic that is not part of the mesh (for a gateway this would be North/South traffic)
   - targetRef:
        kind: meshed # This applies to any traffic as long as it's part of any mesh (in other words has mTLS)
   - targetRef:
        kind: mesh # this applies to traffic from the same mesh as the one of this policy
   - targetRef:
        kind: mesh # this applies to traffic from the mesh "my-mesh"
        name: my-mesh
```

That's the different cases I'm thinking about

Reply:
Can't each of those apply to either only an edge listener or only a cross-mesh listener?

I.e. what those are really doing is actually first selecting which listeners they apply to.
I have the feeling that that is in and of itself confusing and it's kind of using the `from.targetRef` `kind` to narrow the `spec.targetRef`?

_Maybe_ this is OK but another option would just be targeting cross-mesh listeners or edge listeners by using the `spec.targetRef`:

```
spec:
  targetRef:
    kind: MeshGateway
    name: mixed-gateway
from:
  - targetRef:
      kind: Any
```

```
spec:
  targetRef:
    kind: CrossMeshGateway
    name: mixed-gateway
from:
  - targetRef:
      kind: Any # any mesh
  - targetRef:
      kind: Mesh # mesh of gateway
  - targetRef:
      kind: Mesh # some other mesh
      name: my-mesh
```

```
spec:
  targetRef:
    kind: EdgeGateway
    name: mixed-gateway
from:
  - targetRef:
      kind: Any
```

Reply:
> what those are really doing is actually first selecting which listeners they apply to

> using the from.targetRef kind to narrow the spec.targetRef

Can you elaborate on this?

I feel like we already do this on the `outbound` side, we use `to` with `MeshService + name=xyz` to narrow down which `egress` listener we use to route traffic to specific Service.

Reply:
The difference IMO is that the egress listener is much more of an implementation detail than a MeshGateway listener. The gateway listener has an explicit port, it's listed explicitly in the `MeshGateway` and it's something very explicitly addressed by anyone using the gateway. It's a first class object, you can attach a gateway API `HTTPRoute` or `MeshGatewayRoute` to a specific listener.

Reply:
Btw this is maybe more of a side discussion, I added `kind: Any` already. I think I would just prefer to refer to use `spec.targetRef` to select cross-mesh or edge listeners rather than `from.targetRef`.

But it might be a moot point anyway since something has to happen when the user configures:

```
spec:
  targetRef:
    kind: MeshGateway
    name: edge-and-cross-mesh-gateway
from:
  - targetRef:
      kind: Mesh # is this a validation error or does it only apply to cross-mesh listeners
      name: other-mesh 
```

The ship may have sailed on this one for me.